base/MAPL_VerticalMethods.F90

```diff
@@ -248,24 +254,130 @@ subroutine setup_eta_to_pressure(this,regrid_handle,output_grid,rc)
+            if(allocated(this%weight)) deallocate(this%weight)
+            allocate(this%ks(D1,D2,levo),source  = -1)
+            allocate(this%weight(D1,D2,levo),source = 0.0)
+
+            do lev =1, levo
+               pp = flip_sign*this%interp_levels(lev)
+               pb = flip_sign*this%pl3d(:,:,km)
+               do k = km-1, 1,-1 ! levels of input
+                 if(all(pb<pp)) exit
+                 pt = flip_sign*this%pl3d(:,:,k)
+                 where (pp>pt .and. pp<=pb)
+                    this%ks(:,:,lev) = k
+                    this%weight(:,:,lev) = (pb-pp)/(pb-pt)
+                 end where
+                 pb = pt
+               enddo
+            enddo
+            deallocate(this%pl3d) ! not needed any more. release the memory
+
+            ! for edge values
+            if(allocated(this%ks_e))  deallocate(this%ks_e)
+            if(allocated(this%weight_e)) deallocate(this%weight_e)
+            allocate(this%ks_e(D1,D2,levo),source  = -1)
+            allocate(this%weight_e(D1,D2,levo),source = 0.0)
+            km_e = size(this%ple3d,3)
+            this%nedge = km_e
+            allocate(ple3d(D1,D2,km_e))
+            ple3d = this%ple3d 
+            do lev =1, levo
+               pp = flip_sign*this%interp_levels(lev)
+               pb = flip_sign*ple3d(:,:,km_e)
+               !where (pp > pb)
+               !   this%ks_e(:,:,lev) = km_e
+               !   this%weight_e(:,:,lev) = 1.
+               !endwhere
+               do k = km_e-1, 1,-1 ! levels of input
+                 if(all(pb<pp)) exit
+                 pt = flip_sign*ple3d(:,:,k)
+                 where (pp>pt .and. pp<=pb)
+                    this%ks_e(:,:,lev) = k
+                    this%weight_e(:,:,lev) = (pb-pp)/(pb-pt)
+                 end where
+                 pb = pt
+               enddo
+            enddo
+            deallocate(this%ple3d) ! not needed any more. release the memory
+
+            _RETURN(_SUCCESS)
+          end subroutine
+
      end subroutine setup_eta_to_pressure
 
      subroutine regrid_eta_to_pressure(this,ptrin,ptrout,rc)
-        class(verticaldata), intent(inout) :: this
+        class(verticaldata), target, intent(inout) :: this
         real, intent(inout) :: ptrin(:,:,:)
         real, intent(inout) :: ptrout(:,:,:)
         integer, optional, intent(out) :: rc
+        real :: weight
 
         integer :: status
-        integer :: k
+        integer :: i,j,k,lev,levo, D1,D2, km
+        integer, pointer :: ks_(:,:,:)
+        real, pointer    :: weights_(:,:,:)
 
-       do k=1,size(ptrout,3)
-          call vertinterp(ptrout(:,:,k),ptrin,this%interp_levels(k),this%ple3d,this%pl3d,rc=status)
-          _VERIFY(status)
-       end do
-       _RETURN(_SUCCESS)
+        km = size(ptrin,3)
+        if (km == this%nedge -1) then
+           ks_ => this%ks
+           weights_ => this%weight
+        else
+           ks_ => this%ks_e
+           weights_ => this%weight_e
+        endif
```

Comment:
This logic suggests to me that the code further down should be a helper procedure.  The conditional would then either pass (ks,weight) or (ks_e, weight_e) to that helper procedure.    But possibly there are other details that would make this less elegannt.

Reply:
The code further down just uses  ks_ and weighs_.  They don't know if it is center or edge any more.